pkg/kubelet/kuberuntime/kuberuntime_logs_test.go

```diff
@@ -0,0 +1,191 @@
+		opts := newLogOptions(test.apiOpts, timestamp.Time)
+		assert.Equal(t, test.expect, opts)
+	}
+}
+
+func TestParseLog(t *testing.T) {
+	timestamp, err := time.Parse(timeFormat, "2016-10-20T18:39:20.57606443Z")
+	assert.NoError(t, err)
+	msg := &logMessage{}
+	for c, test := range []struct {
+		line string
+		msg  *logMessage
+		err  bool
+	}{
+		{ // Docker log format stdout
+			line: `{"log":"docker stdout test log","stream":"stdout","time":"2016-10-20T18:39:20.57606443Z"}` + "\n",
+			msg: &logMessage{
+				timestamp: timestamp,
+				stream:    stdoutType,
+				log:       []byte("docker stdout test log"),
+			},
+		},
+		{ // Docker log format stderr
+			line: `{"log":"docker stderr test log","stream":"stderr","time":"2016-10-20T18:39:20.57606443Z"}` + "\n",
+			msg: &logMessage{
+				timestamp: timestamp,
+				stream:    stderrType,
+				log:       []byte("docker stderr test log"),
+			},
+		},
+		{ // CRI log format stdout
+			line: "2016-10-20T18:39:20.57606443Z stdout cri stdout test log\n",
+			msg: &logMessage{
+				timestamp: timestamp,
+				stream:    stdoutType,
+				log:       []byte("cri stdout test log\n"),
+			},
+		},
+		{ // CRI log format stderr
+			line: "2016-10-20T18:39:20.57606443Z stderr cri stderr test log\n",
+			msg: &logMessage{
+				timestamp: timestamp,
+				stream:    stderrType,
+				log:       []byte("cri stderr test log\n"),
+			},
+		},
+		{ // Unsupported Log format
+			line: "unsupported log format test log\n",
+			msg:  &logMessage{},
+			err:  true,
+		},
+	} {
+		t.Logf("TestCase #%d: %+v", c, test)
+		parse, err := getParseFunc([]byte(test.line))
+		if test.err {
+			assert.Error(t, err)
+			continue
+		}
+		assert.NoError(t, err)
+		err = parse([]byte(test.line), msg)
+		assert.NoError(t, err)
+		assert.Equal(t, test.msg, msg)
+	}
+}
+
+func TestWriteLogs(t *testing.T) {
+	timestamp := time.Unix(1234, 4321)
+	log := "abcdefg\n"
+
+	for c, test := range []struct {
+		stream       streamType
+		since        time.Time
+		timestamp    bool
+		expectStdout string
+		expectStderr string
+	}{
+		{ // stderr log
+			stream:       stderrType,
+			expectStderr: log,
+		},
+		{ // stdout log
+			stream:       stdoutType,
+			expectStdout: log,
+		},
+		{ // since is after timestamp
+			stream: stdoutType,
+			since:  timestamp.Add(1 * time.Second),
+		},
+		{ // timestamp enabled
+			stream:       stderrType,
+			timestamp:    true,
+			expectStderr: timestamp.Format(timeFormat) + " " + log,
+		},
+	} {
+		t.Logf("TestCase #%d: %+v", c, test)
+		msg := &logMessage{
+			timestamp: timestamp,
+			stream:    test.stream,
+			log:       []byte(log),
+		}
+		stdoutBuf := bytes.NewBuffer(nil)
+		stderrBuf := bytes.NewBuffer(nil)
+		err := writeLogs(msg, stdoutBuf, stderrBuf, &logOptions{since: test.since, timestamp: test.timestamp})
+		assert.NoError(t, err)
+		assert.Equal(t, test.expectStdout, stdoutBuf.String())
+		assert.Equal(t, test.expectStderr, stderrBuf.String())
+	}
+}
+
+func TestTail(t *testing.T) {
```

Comment:
Will do. BTW, I think the function will only use around `blockSize=1024b` memory, because we are always using the same buf - `buf := make([]byte, blockSize)`. It should be much better than docker's implementation, which read all logs into the memory https://github.com/docker/docker/blob/master/pkg/tailfile/tailfile.go#L20.